lib/src/test/resources/compileContracts.sh

```diff
@@ -1,47 +1,134 @@
 #!/bin/bash
 
-pushd "$(dirname "$0")" || exit
-mkdir -p "compiled_v0"
+REPO_ROOT=$(dirname "$0")/../../../..
+
+V0_CONTRACT_PATH="src_v0"
+V1_CONTRACT_PATH="src_v1"
+V2_CONTRACT_PATH="src_v2"
+
+V0_ARTIFACT_PATH="compiled_v0"
+V1_ARTIFACT_PATH="compiled_v1"
+V2_ARTIFACT_PATH="compiled_v2"
+
+V1_COMPILER_VERSION="1.1.1"
+V2_COMPILER_VERSION="2.2.0"
+
+V1_COMPILER_BIN_PATH="compilers/v1"
+V2_COMPILER_BIN_PATH="compilers/v2"
+
+build_cairo_compilers() {
+  local VERSION=$1
+  local OUT_DIR=$2
+
+  local VERSION_SHORT=${VERSION:0:1}
+
+  if [ ! -d "$REPO_ROOT/cairo$VERSION_SHORT" ]; then
+    echo "Cloning cairo repo..."
+    git clone -q https://github.com/starkware-libs/cairo.git "$REPO_ROOT/cairo$VERSION_SHORT"
+    pushd "$REPO_ROOT/cairo$VERSION_SHORT" || exit 1
+    git checkout -q "v$VERSION"
+    popd
+  else
+    echo "Found existing cairo repo, skipping cloning."
+  fi
+
+  if ! find "$REPO_ROOT/cairo$VERSION_SHORT/target/release" -name starknet-compile -type f > /dev/null || ! find "$REPO_ROOT/cairo$VERSION_SHORT/target/release" -name starknet-sierra-compile -type f > /dev/null; then
+    if ! which cargo >/dev/null; then
+      echo "Installing rust..."
+      curl -sSf https://sh.rustup.rs | sh -s -- -y
+    fi
+    echo "Building starknet compiler"
+    pushd "$REPO_ROOT/cairo$VERSION_SHORT" || exit 1
+    cargo build --release
+    cargo run --bin starknet-compile -- --version
+    cargo run --bin starknet-sierra-compile -- --version
+    popd
+  else
+    echo "Found existing binaries, skipping compilation."
+  fi
+
+  rm -r "$(dirname "$0")/$OUT_DIR/cairo/bin/" || true
+  rm -r "$(dirname "$0")/$OUT_DIR/cairo/corelib/" || true
+
+  mkdir -p "$(dirname "$0")/$OUT_DIR/cairo/bin/"
+  mkdir -p "$(dirname "$0")/$OUT_DIR/cairo/corelib/"
+
+  echo "Copying binaries..."
+  rsync -aW "$REPO_ROOT/cairo$VERSION_SHORT/target/release/" "$(dirname "$0")/$OUT_DIR/cairo/bin/"
+  echo "Copying corelib..."
+  rsync -aW "$REPO_ROOT/cairo$VERSION_SHORT/corelib/" "$(dirname "$0")/$OUT_DIR/cairo/corelib/"
+}
+
+fetch_compilers() {
+  local VERSION=$1
+  local OUT_DIR=$2
+
+  local VERSION_SHORT=${VERSION:0:1}
+
+  echo "Fetching compiler v$VERSION binaries..."
+
+  mkdir -p "$OUT_DIR"
+
+  OS=$(uname)
+  ARCH=$(uname -m)
+
+  if [ "$OS" == "Linux" -a "$ARCH" == "x86_64" ]; then
+    curl -LsSf "https://github.com/starkware-libs/cairo/releases/download/v${VERSION}/release-x86_64-unknown-linux-musl.tar.gz" | tar -xz -C "$OUT_DIR" || exit 1
+  elif [ "$OS" == "Darwin" -a "$ARCH" == "arm64" ]; then
+    curl -LsSf "https://github.com/starkware-libs/cairo/releases/download/v${VERSION}/release-aarch64-apple-darwin.tar" | tar -x -C "$OUT_DIR" || exit 1
+  elif [ "$OS" == "Darwin" -a "$ARCH" == "x86_64" ]; then
+    build_cairo_compilers "$VERSION" "$OUT_DIR"
```

Comment:
I'd either change it to be more general - not only for macs, or remove it altogether. Would probably go with the latter, as I don't think we should add more unnecessary code for us to maintain.

If a binary for system/arch is not available to download we could allow to get the path to compiler from env/variable passed to script or throw an error.

Reply:
Are there any other platforms? This is one is popular, used by us on CI, is compatible with all of our dependencies, and effortless to "maintain". Also, it seems to me that making it more general would hide the fact that of all of the other platforms only macOS x64 is supported. 

I mentioned as to why approach with path seems problematic to me [in other comment](https://github.com/software-mansion/starknet-jvm/pull/285#discussion_r1324164073).

Reply:
>Are there any other platforms? 

Loads of :D 

>Thi one is popular, used by us on CI

but we will checkout and compile cairo on CI, because we have to, right? So I'd rather do everything that is required there. Especially so that this platform is dying, and we plan to move away from it to apple sillicon on CI eventually, right?

> Also, it seems to me that making it more general would hide the fact that of all of the other platforms only macOS x64 is supported.

supported platforms != platforms you can use to contribute to starknet-jvm from

Reply:
> supported platforms != platforms you can use to contribute to starknet-jvm from

I see, my bad using those interchangeably.

> but we will checkout and compile cairo on CI, because we have to, right? So I'd rather do everything that is required there.

Do you have moving the entire block with compiling contracts to CI? Both script and checks might get more complicated it seems to me. Certainly in such scenario no contribution guidelines can be added.

Reply:
No biggie :) 

I suppose we could just compile the compiler on CI (macos only), and then allow for scripts to use custom compilers (either by env variable or an argument passed to script, former probably makes more sense). I guess this will make sure that everything works for everyone, and is not very complicated. Contribution guide would then have a general info on where to get cairo code from, how to compile it and how to pass it to script.

Sounds good?

Reply:
Can't wait for us to finally migrate to devnet-rs and scarb :) every problem we discussed will be fixed by this